pipeline/translate/translate_ctranslate2.py

```diff
@@ -0,0 +1,184 @@
+            return int(value)
+        raise ValueError(f'Expected "{key}" to be of a type "{type}" in the decoder.yml config')
+
+
+def write_single_translation(
+    _index: int, tokenizer_trg: spm.SentencePieceProcessor, result: Any, outfile: TextIO
+):
+    """
+    Just write each single translation to a new line. If beam search was used all the other
+    beam results are discarded.
+    """
+    line = tokenizer_trg.decode(result.hypotheses[0])
+    outfile.write(line)
+    outfile.write("\n")
+
+
+def write_nbest_translations(
+    index: int, tokenizer_trg: spm.SentencePieceProcessor, result: Any, outfile: TextIO
+):
+    """
+    Match Marian's way of writing out nbest translations. For example, with a beam-size of 2 and
+    collection nbest translations:
+
+    0 ||| Translation attempt
+    0 ||| An attempt at translation
+    1 ||| The quick brown fox jumped
+    1 ||| The brown fox quickly jumped
+    ...
+    """
+    for hypothesis in result.hypotheses:
+        line = tokenizer_trg.decode(hypothesis)
+        outfile.write(f"{index} ||| {line}\n")
+
+
+def translate_with_ctranslate2(
+    input_zst: Path,
+    artifacts: Path,
+    extra_marian_args: list[str],
+    models_globs: list[str],
+    is_nbest: bool,
+    vocab: list[str],
+    device: str,
+    device_index: list[int],
+) -> None:
+    model = get_model(models_globs)
+    postfix = "nbest" if is_nbest else "out"
+
+    tokenizer_src = spm.SentencePieceProcessor(vocab[0])
+    if len(vocab) == 1:
+        tokenizer_trg = tokenizer_src
+    else:
+        tokenizer_trg = spm.SentencePieceProcessor(vocab[1])
+
+    if extra_marian_args and extra_marian_args[0] != "--":
+        logger.error(" ".join(extra_marian_args))
+        raise Exception("Expected the extra marian args to be after a --")
+
+    decoder_config = DecoderConfig(extra_marian_args[1:])
+
+    ctranslate2_model_dir = model.parent / f"{Path(model).stem}"
+    logger.info("Converting the Marian model to Ctranslate2:")
+    logger.info(model)
+    logger.info("Outputing model to:")
+    logger.info(ctranslate2_model_dir)
+
+    converter = MarianConverter(model, vocab)
+    converter.convert(ctranslate2_model_dir, quantization=decoder_config.precision)
+
+    if device == "gpu":
+        translator = ctranslate2.Translator(
+            str(ctranslate2_model_dir), device="cuda", device_index=device_index
+        )
+    else:
+        translator = ctranslate2.Translator(str(ctranslate2_model_dir), device="cpu")
+
+    logger.info("Loading model")
+    translator.load_model()
+    logger.info("Model loaded")
+
+    output_zst = artifacts / f"{input_zst.stem}.{postfix}.zst"
+
+    num_hypotheses = 1
+    write_translation = write_single_translation
+    if is_nbest:
+        num_hypotheses = decoder_config.beam_size
+        write_translation = write_nbest_translations
+
+    def tokenize(line):
+        return tokenizer_src.Encode(line.strip(), out_type=str)
```

Comment:
Are we sure we want to strip whitespace on both side? Or do we want to just remove the new line symbol?

Reply:
Good catch, yeah it'll be better to just remove the newline.